clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```
@@ -116,9 +116,9 @@
     public void seekToEnd(Collection<TopicPartition> partitions);
 
     /**
-     * @see KafkaConsumer#position(TopicPartition)
+     * @see KafkaConsumer#position(TopicPartition, long)
      */
-    public long position(TopicPartition partition);
+    public long position(TopicPartition partition, long timeout);
```

API changes will call for a KIP I'm afraid. We'll also want to ensure that we preserve the old method for compatibility.

---

Thanks for the review @hachikuji maybe we can avoid the public API change by introducing a new config value, or I am not 100 percent sure about this but maybe we can avoid that timeout thing at all and find a new way to handle this problem. (This needs more investigation from my side:))

---

cc @huxihx

---

Whether it's a new config or a new API, we'll still need a KIP. If we go that far, we'll want to consider the other methods which probably suffer the same problem (e.g. `seekToBeginning`). A couple options that we have thrown around are the following:

1. Introduce a `max.block.ms` config like the producer has.
2. Overload the methods with timeouts.
3. Use the request timeout.

Not sure the best way to go here to be honest. Maybe for `position()`, we should just raise a `UnknownTopicException` instead of retrying.

---

Actually when I was discussing with @huxihx on the JIRA itself what I was thinking is the following:

1. The call traces of `Fetcher#resetOffsets` which use `INT_MAX` are from two public APIs: 

```
Fetcher#resetOffsets -> 
Fetcher#updateFetchPositions / Fetcher#resetOffsetsIfNeeded ->
KafkaConsumer#updateFetchPositions ->
KafkaConsumer#poll(timeout) / KafkaConsumer#position()
```

So we should pass in a `remaining` in the internal call from `KafkaConsumer#updateFetchPositions` all the way down to `Fetcher#resetOffsets` to replace `INT_MAX`; 

As for the public APIs, `poll` already have a timeout, we just need to let its internal `pollOnce` to pass in the updated `remaining` to `updateFetchPositions`, for `position` since it is defined as a non-blocking call, we should pass `0` to `updateFetchPositions` and if the call fails to return the offset in a single trial, it means no offset is cached locally and the remote offset request failes as well, we can throw `InvalidOffsetException` directly.

By doing this we do not need a KIP. Admittedly it may not be optimal to treat the non-blocking `position()` call like this, but this is to adherent with its semantics; if we do want to change its public behavior then I'd prefer option 1) from @hachikuji to modify the semantics of "non-blocking" calls of consumer as we did for producer.

---

Passing zero inside the position method I think is not a good idea because we are going to introduce a lot of flaky test. Sometimes the future will still running when we are checking.
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java#L458

---

OK. I think I understand the problem a little better after reading @guozhangwang's comment on the JIRA. I'm thinking it might be sufficient to handle this case specially. When automatic assignment is used, there really should be no reason we block on partitions that we know don't exist. From a high level, I'm wondering if we can consult the local metadata cache and if we see that the partition number is greater than the number of partitions, then 1) if we're inside `position()`, we raise an exception, and 2) if we're inside `poll()`, we effectively pause that partition and request another rebalance.

---

The issue is that local cache may not contain this topic metadata yet or not up-to-date, and that's why we may want to send an `MetadataRequest` in these two function calls.

---

Yes I think after we update metadata, we should do a sanity check on the number of partitions. If we have stale information (i.e. the topic's state prior to deletion), then the partition will still exist and we can retry. But once we have updated state, we will see that the topic has a fewer number of partitions and we can raise an exception. Then how we handle it depends on the context.

---

I think the root cause of the hanging is that, the `list offset` response returns empty indicating the broker did not know about this topic, and the client will hence retry forever. So what I suggest is that just making the metadata refresh once in `position()`. As for the flaky unit tests @baluchicken , we can use waitForConditions in those test rather than block waiting until it succeeds.

---

Sorry I don't have time to deal with this, I will continue to work on this in Monday:)